### PR TITLE
#95 First shot for implementing Focal Point Cropping

### PR DESCRIPTION
**Pull request by duergner**
There are still tests missing and the way it has been implemented required changing all other filters to inject the basic Skipper FilterContext. This should most probably be refactored.

refs #95

### CODE REVIEW

**Comment by duergner** [comment]
👎

**Comment by danpersa** [comment]
Please modify [this filter](https://github.com/zalando-stups/skrop/blob/master/filters/transformFromQueryParams.go) as well to support this new feature. The free transform should support all of the new features we add.

**Comment by duergner** [comment]
Yeah thought about this as well. The downside I see with this filter is that it required to use a query param instead of path variables.

File: filters/cropbyfocalpoint.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+// NewCropByFocalPoint creates a new filter of this type
+func NewCropByFocalPoint() filters.Spec {
+	return &cropByFocalPoint{}
+}
+
+func (f *cropByFocalPoint) Name() string {
+	return CropByFocalPointName
+}
+
+func (f *cropByFocalPoint) CreateOptions(imageContext *ImageFilterContext, filterContext filters.FilterContext) (*bimg.Options, error) {
```
Comment (danpersa) [inline filters/cropbyfocalpoint.go]: The filter context should not be passed here. Instead please extend [this method](https://github.com/zalando-stups/skrop/blob/59b76fe5c1e088ee131d62c37199566f2cb81305/filters/imagefilter.go#L93) to add the required information from the `FilterContext` to the `ImageFilterContext`

File: filters/imagefilter.go [hunk 2]
```diff
@@ -80,8 +80,17 @@ func errorResponse() *http.Response {
...
 func buildParameters(ctx filters.FilterContext, image *bimg.Image) *ImageFilterContext {
 	parameters := map[string][]string(nil)
+
 	if ctx != nil {
 		parameters = ctx.Request().URL.Query()
+
+		focalPointX := ctx.PathParam("focalPointX");
+		focalPointY := ctx.PathParam("focalPointY");
+
+		if len(focalPointX) > 0 && len(focalPointY) > 0 {
```
Comment (danpersa) [inline filters/imagefilter.go]: I would not add the `PathParams` to the parameters, I would rather add all of the path params to the `ImageFilterContext`, something like this:

```
&ImageFilterContext{
 		Image:      image,
.....
                PathParams: pathParams
}
```

The reason is that with your approach, every time a new filter adds a new parameter we need to modify this method, which breaks the [Open Closed Principle](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle). As soon as the path parameters are part of the ImageFilterContext, each of the filters can access them.

**Comment by danpersa** [comment]
Please don't forget to add tests (the build will fail otherwise) and update the README (there is a section there where we mention all of the filters)

File: filters/imagefilter.go (hunk 2 above)
Comment (duergner) [reply filters/imagefilter.go]: @danpersa the problem is that Skipper currently does not allow to get all PathParams. You need to access them by name.

I can move them out of the parameters but unless there is a change in Skipper I don't see a way to add all PathParams without knowing which exist.

**Comment by duergner** [comment]
@danpersa can you have another look. I added both tests for the new filter as well as an entry to the README file.

File: filters/cropbyfocalpoint.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
+package filters
+
+import (
+	log "github.com/sirupsen/logrus"
+	"github.com/zalando-stups/skrop/parse"
+	"github.com/zalando/skipper/filters"
+	"gopkg.in/h2non/bimg.v1"
+	"strconv"
+	)
```
Comment (addityasingh) [inline filters/cropbyfocalpoint.go]: nitpicking: I am not sure if gofmt is working correctly here. The `)` should be at start of the line

File: filters/cropbyfocalpoint.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+func NewCropByFocalPoint() filters.Spec {
+	return &cropByFocalPoint{}
+}
+
+func (f *cropByFocalPoint) Name() string {
+	return CropByFocalPointName
+}
+
+func (f *cropByFocalPoint) CreateOptions(imageContext *ImageFilterContext) (*bimg.Options, error) {
+	log.Debug("Create options for crop by focal point ", f)
```
Comment (addityasingh) [inline filters/cropbyfocalpoint.go]: Please remove the debug line before merge.

File: filters/cropbyfocalpoint_test.go [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+	assert.Equal(t, "cropByFocalPoint", name)
+}
+
+func TestCropByFocalPoint_Name(t *testing.T) {
+	c := cropByFocalPoint{}
+	assert.Equal(t, "cropByFocalPoint", c.Name())
+}
+
+func TestCropByFocalPoint_CreateOptions(t *testing.T) {
+	c := cropByFocalPoint{targetX: 0.5, targetY: 0.5, aspectRatio: 1.5}
```
Comment (addityasingh) [inline filters/cropbyfocalpoint_test.go]: Probably these repeating tests could be moved to a TableDriven tests?

**Comment by addityasingh** [comment]
@duergner I have added minor comments. Please check if they make sense

File: filters/cropbyfocalpoint_test.go (hunk 5 above)
Comment (duergner) [reply filters/cropbyfocalpoint_test.go]: Given my very limited Go experience I wouldn't disagree if you suggest that.

I'll have a look on how to do this.

**Comment by duergner** [comment]
@addityasingh I did the two small fixes. Will have a look into the restructure to TableDriven tests.

File: filters/transformFromQueryParams.go [hunk 6]
```diff
@@ -30,6 +31,19 @@ func (t *transformFromQueryParams) CreateOptions(ctx *ImageFilterContext) (*bimg
...
+		if !ok {
+			return &bimg.Options{}, nil
+		}
+
+		params = strings.Split(params[0], ",")
+		if len(params) != 5 {
+			return &bimg.Options{}, nil
+		}
+
+		// TODO do focal point crop
```
Comment (danpersa) [inline filters/transformFromQueryParams.go]: @duergner can you please open a ticket for this part, so that we don't forget about it?

**Comment by danpersa** [comment]
👍

File: filters/cropbyfocalpoint_test.go (hunk 5 above)
Comment (addityasingh) [reply filters/cropbyfocalpoint_test.go]: Great. I am new too, but would love to be of help on this :)

File: filters/transformFromQueryParams.go (hunk 6 above)
Comment (duergner) [reply filters/transformFromQueryParams.go]: @danpersa done: https://github.com/zalando-stups/skrop/issues/99

**Pull request merged**